### PR TITLE
6940 vscode setup

### PR DESCRIPTION
**Pull request by saintmatthieu**
Resolves: #6940

 Help developers who want to use Visual Studio or VSCode to easily get setup.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -4,7 +4,7 @@ root = true
 
 [*]
 indent_style = space
-indent_size = 3
+indent_size = 4
```
Comment (LWinterberg) [inline .editorconfig:7]: good lord finally

File: .vscode/c_cpp_properties.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{
+   "configurations": [
+      {
+         "name": "Win32",
```
Comment (igorkorsukov) [inline .vscode/c_cpp_properties.json:4]: It should probably just be `Win`

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
+    "cmake.generator": "Ninja",
+    "cmake.configureArgs": [
+        "-DCMAKE_INSTALL_PREFIX=${workspaceFolder}/build/install",
```
Comment (kryksyh) [inline .vscode/settings.json:4]: while we are at it, would you mind adding CMakeUserPresets.json?

 see https://github.com/audacity/audacity-private-docs/blob/main/build/CMake%20Presets.md for AU3
 
 Or we could do that within another PR

File: developer.bat [hunk 4]
```diff
@@ -0,0 +1,17 @@
+IF NOT EXIST build mkdir build
+
+cd build
+
+cmake -DCMAKE_INSTALL_PREFIX=install -G "Visual Studio 16 2019" -DDEVELOPER=TRUE ..
+
+cmake --build . --config Debug --target install
```
Comment (igorkorsukov) [inline developer.bat:7]: I think there is no need to perform the building and installation itself, it is enough to just configure the project to generate .sln

File: developer.bat [hunk 5]
```diff
@@ -0,0 +1,17 @@
+IF NOT EXIST build mkdir build
+
+cd build
+
+cmake -DCMAKE_INSTALL_PREFIX=install -G "Visual Studio 16 2019" -DDEVELOPER=TRUE ..
```
Comment (kryksyh) [inline developer.bat:5]: Should we hardcode VS version? I'm using 2022

File: src/app/CMakeLists.txt [hunk 6]
```diff
@@ -273,6 +273,20 @@ if (OS_IS_WIN)
               USE_SOURCE_PERMISSIONS
           )
 
+    if (MSVC AND DEVELOPER)
```
Comment (igorkorsukov) [inline src/app/CMakeLists.txt:276]: DEVELOPER - I think we need to think of another name, it’s not clear what it is and where it comes from 
I think we need to check the name of the cmake generator 
Or something like this: VS_CMAKE_GENERATOR

File: .vscode/c_cpp_properties.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+         "includePath": [
+            "${workspaceFolder}/**",
+            "${QTDIR}/include/**"
+         ],
+         "defines": [
+            "_DEBUG",
+            "UNICODE",
+            "_UNICODE"
+         ],
+         "windowsSdkVersion": "10.0.22621.0",
```
Comment (kryksyh) [inline .vscode/c_cpp_properties.json:14]: Do we need it hardcoded here? I believe that it can be detected automatically.

File: .vscode/c_cpp_properties.json [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+         "compilerPath": "cl.exe",
+         "cStandard": "c17",
+         "cppStandard": "c++17",
+         "intelliSenseMode": "msvc-x64",
+         "configurationProvider": "ms-vscode.cmake-tools"
+      }
+   ],
+   "version": 4
+}
```
Comment (kryksyh) [inline .vscode/c_cpp_properties.json:23]: Some files are missing newlines

File: src/app/CMakeLists.txt (hunk 6 above)
Comment (igorkorsukov) [reply src/app/CMakeLists.txt:276]: I see that this is also used for VSCode, then we still need to think about a good name.

File: .vscode/c_cpp_properties.json (hunk 8 above)
Comment (kryksyh) [reply .vscode/c_cpp_properties.json:23]: does vscode not obey "insert_final_newline" for .editorconfig?

File: .vscode/extensions.json [hunk 9]
```diff
@@ -0,0 +1,7 @@
+{
+    "recommendations": [
+        "ms-vscode.cmake-tools",
```
Comment (kryksyh) [inline .vscode/extensions.json:3]: Shouldn't we add "SynthWave '84" them with glow enabled to the list of recommendations? (jk)

File: .vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        "-DDEVELOPER=TRUE"
+        // "-DAU_MODULE_EFFECTS_VST=ON"
+    ],
+    "cmake.environment": {
+        "QTDIR": "C:/Qt/6.2.0/msvc2019_64"
+    },
+    "qt-official.qmlls.customExePath": "C:/Qt/Tools/QtDesignStudio/qt6_design_studio_reduced_version/bin/qmlls.exe",
```
Comment (igorkorsukov) [inline .vscode/settings.json:11]: better to use QTDIR

File: developer.bat [hunk 11]
```diff
@@ -0,0 +1,17 @@
+IF NOT EXIST build mkdir build
```
Comment (igorkorsukov) [inline developer.bat:1]: developer.bat - I think it would be better to call it something like `make_sln.bat` or `generate_sln.bat`

File: .vscode/c_cpp_properties.json (hunk 2 above)
Comment (igorkorsukov) [reply .vscode/c_cpp_properties.json:4]: Found that this is an internal name in VSСode

File: .vscode/c_cpp_properties.json (hunk 7 above)
Comment (igorkorsukov) [reply .vscode/c_cpp_properties.json:14]: As far as I can see, this is (c_cpp_properties.json) an automatically created file, it is better to add it to gitignore.

File: .vscode/launch.json [hunk 12]
```diff
@@ -0,0 +1,25 @@
+{
+    "configurations": [
+        {
+            "name": "run audacity",
+            "type": "cppvsdbg",
+            "request": "launch",
+            "program": "${workspaceFolder}/build/install/bin/audacity.exe",
+            "preLaunchTask": "build audacity",
```
Comment (igorkorsukov) [inline .vscode/launch.json:8]: Probably there should be not a build here, but an install.
@kryksyh  suggested that we can just do `CMake: Install`, and not add a task file.

File: .vscode/launch.json [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+            "cwd": "${workspaceFolder}",
+            "visualizerFile": "${workspaceFolder}/qt6.natvis",
+            "sourceFileMap": {
+                "C:/Users/qt/work/qt": "${QTDIR}/../Src"
+            },
+            "windows": {
+                "environment": [
+                    {
+                        "name": "PATH",
+                        "value": "${env:PATH};${command:qt-cpp.kitDirectory}${/}bin"
```
Comment (igorkorsukov) [inline .vscode/launch.json:19]: It is better to simply write in the instructions (and when trying to configure it, it is indicated) that us need to add Qt to the PATH

File: .vscode/settings.json (hunk 3 above)
Comment (saintmatthieu) [reply .vscode/settings.json:4]: You see, I missed that. Will consider it for this PR.

File: developer.bat (hunk 4 above)
Comment (saintmatthieu) [reply developer.bat:7]: The proposed approach in this PR is to add an audacity-target post-build step to copy the executable to the install directory. If you just build audacity without building install first, it will copy it, it will be alone in its `install/bin`.

File: developer.bat (hunk 5 above)
Comment (saintmatthieu) [reply developer.bat:5]: Can we ensure a VS generator without specifying the version? I'd gladly remove this flag, but if I do, I don't know what ensures that it will pick a visual studio generator.

Comment (saintmatthieu) [reply developer.bat:5]: The only solution I've found to that is a bit complicated, but maybe worth using:
```bat
REM Detect the highest available Visual Studio version
for /f "tokens=*" %%i in ('"C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe" -latest -products * -requires Microsoft.Component.MSBuild -property installationVersion') do set VS_VERSION=%%i

REM The `VS_VERSION` will be something like 17.11.35312.102 ; keep only the major version
set VS_VERSION=%VS_VERSION:~0,2%

if %VS_VERSION%==16 (
    set GENERATOR="Visual Studio 16 2019"
) else if %VS_VERSION%==17 (
    set GENERATOR="Visual Studio 17 2022"
) else (
    echo No supported Visual Studio version found.
    exit /b 1
)

echo Using %GENERATOR%
```

File: src/app/CMakeLists.txt (hunk 6 above)
Comment (saintmatthieu) [reply src/app/CMakeLists.txt:276]: I just realize that the `MSVC` check is sufficient for this part of the configuration (configuring the debugging command correctly). This then leaves the `DEVELOPER` flag to adding the post-build step to copy the executable to the install directory, which we can still discuss if we want this at all.

File: .vscode/c_cpp_properties.json (hunk 7 above)
Comment (saintmatthieu) [reply .vscode/c_cpp_properties.json:14]: > Do we need it hardcoded here? I believe that it can be detected automatically.

Probably.

Comment (saintmatthieu) [reply .vscode/c_cpp_properties.json:14]: > As far as I can see, this is (c_cpp_properties.json) an automatically created file, it is better to add it to gitignore.

The `includePath`, adding the QTDIR, isn't default and it's something I enjoy.
Once the file is created, I'm not sure it gets modified. I suggest we give it a try and re-assess if we see that it does and it becomes a nuisance.

File: .vscode/c_cpp_properties.json (hunk 8 above)
Comment (saintmatthieu) [reply .vscode/c_cpp_properties.json:23]: We have this:
```
[*.json]
insert_final_newline = unset
```
Are the missing newlines an issue, though?

File: .vscode/extensions.json (hunk 9 above)
Comment (saintmatthieu) [reply .vscode/extensions.json:3]: We should recommend the uncrustify extension, though. Will add.

File: src/app/CMakeLists.txt (hunk 6 above)
Comment (igorkorsukov) [reply src/app/CMakeLists.txt:276]: If we decide to keep the copy step, let's name the flag something like: `OPTIMIZED_DEPLOY`

File: .vscode/extensions.json (hunk 9 above)
Comment (saintmatthieu) [reply .vscode/extensions.json:3]: > Shouldn't we add "SynthWave '84" theme with glow enabled to the list of recommendations? (jk)

I don't know. Does it work well with Doki-Theme?

File: .vscode/settings.json (hunk 10 above)
Comment (saintmatthieu) [reply .vscode/settings.json:11]: Of course.

File: .vscode/c_cpp_properties.json (hunk 7 above)
Comment (igorkorsukov) [reply .vscode/c_cpp_properties.json:14]: I believe we can add Qt to the include path in our cmake files.
Although it's strange that I didn't need it in my example...
Maybe because I have Qt in the PATH

File: .vscode/settings.json [hunk 14]
```diff
@@ -0,0 +1,30 @@
+{
+    "cmake.generator": "Ninja",
+    "cmake.configureArgs": [
+        "-DCMAKE_INSTALL_PREFIX=${workspaceFolder}/build/install",
+        "-DDEVELOPER=TRUE"
+        // "-DAU_MODULE_EFFECTS_VST=ON"
+    ],
+    "cmake.environment": {
+        "QTDIR": "C:/Qt/6.2.0/msvc2019_64"
```
[the remark this answers is not in the document]
Comment (saintmatthieu) [reply .vscode/settings.json:9]: In fact, I think we can get rid of this entry altogether. I expect cmake will query the QTDIR env variable by default, and we do expect contributors to have this set.

File: developer.bat (hunk 11 above)
Comment (saintmatthieu) [reply developer.bat:1]: I agree. It's nostalgia: my previous company had it named that way. I also like the idea of a developer 🦇

Comment (saintmatthieu) [reply developer.bat:1]: generate_sln sounds good.

File: .vscode/launch.json (hunk 12 above)
Comment (saintmatthieu) [reply .vscode/launch.json:8]: > Probably there should be not a build here, but an install
This works because of the post-build task in the cmakelists that gets the executable copied to the install dir.
Having `install` as pre-launch task has for consequence that each time you launch after having modified the source code, the install target gets executed and, even if it's not the first time, just the check that everything (except audacity.exe) is up-to-date takes at least a couple of seconds.
On the other hand, for fast onboarding, what you suggest would be easier: the user wouldn't have to build the install target manually once.
But we can propose two launch tasks: the one you suggest and make it the default one, and for power users, the present one.

Comment (saintmatthieu) [reply .vscode/launch.json:8]: > even if it's not the first time, just the check that everything (except audacity.exe) is up-to-date takes at least a couple of seconds

I'm confused. Now it's super fast for me even with install. Ok for your proposal, then.

File: src/app/CMakeLists.txt (hunk 6 above)
Comment (saintmatthieu) [reply src/app/CMakeLists.txt:276]: Sorry, no, it still is needed for Visual Studio, which doesn't have pre-launch tasks AFAIK.

Comment (saintmatthieu) [reply src/app/CMakeLists.txt:276]: ... but then it's not so much a matter of optimizing deployment anymore. `COPY_TO_INSTALL` ? Doesn't say everything but it gives a good clue.

Comment (saintmatthieu) [reply src/app/CMakeLists.txt:276]: ... or maybe we can just add that post-build step just if `MSVC`. Won't hurt CI anyway, as it uses Ninja. Then no need for that extra flag.

File: BUILDING.md [hunk 15]
```diff
@@ -39,15 +39,43 @@ Git, CMake, Ninja, Package manager, Compiler and Qt should all be added to the P
...
 ```
 # inside the audacity source:
 cmake -S . -B build/ [options]  # configure (first build only)
 cmake --build build/            # build (every build)
 cmake --install build/          # install (every successful build)
 ```
+
+### With Visual Studio (Windows only)
+
+Double-click the `developer.bat` script in the root of the repository. This will generate a Visual Studio solution in the `build` directory and build the `install` target. Open the generated solution (./build/audacity.sln) in Visual Studio and press F5 to run Audacity.
```
Comment (igorkorsukov) [inline BUILDING.md:61]: `developer.bat` -> `generate_sln.bat`

**Review by igorkorsukov — APPROVED** [review]

File: developer.bat (hunk 5 above)
Comment (kryksyh) [reply developer.bat]: I see, we can give it a shot. One thing though, it should be `%programfiles(x86)%` instead of `C:\Program Files (x86)`

File: .vscode/c_cpp_properties.json (hunk 8 above)
Comment (kryksyh) [reply .vscode/c_cpp_properties.json:22]: > Are the missing newlines an issue, though?

It's not that important, it's just the way it's done historically. And without a newline, some command line tools can misbehave. For example if you `cat file` the prompt will be printed on the same line as the last line from the file

image

File: src/app/CMakeLists.txt (hunk 6 above)
Comment (igorkorsukov) [reply src/app/CMakeLists.txt]: `COPY_TO_INSTALL` - acceptable for me

File: .vscode/extensions.json (hunk 9 above)
Comment (kryksyh) [reply .vscode/extensions.json:3]: Unfortunately not :(

**Review by kryksyh — APPROVED** [review]
